neptune-core/src/models/state/archival_state.rs

```diff
@@ -868,6 +863,87 @@ impl ArchivalState {
         Ok(Some(block))
     }
 
+    /// Returns a [`HashMap`] of [`AdditionRecord`] to [`Option`] of AOCL leaf
+    /// index (`u64`) for all outputs in a given block. If the block is not
+    /// canonical, the indices are all `None`, and conversely, if the block is
+    /// canonical then the indices point into the current mutator set AOCL.
+    /// If the block does not live in the archival state, return `None`.
+    ///
+    /// # Panics
+    ///
+    ///  - If the block is not canonical, was stored, and reading it from disk
+    ///    fails.
+    ///  - If the block is not canonical, was stored, and is invalid.
+    pub(crate) async fn get_addition_record_indices_for_block(
+        &self,
+        block_digest: Digest,
+    ) -> Option<HashMap<AdditionRecord, Option<u64>>> {
+        let maybe_block_record = self.get_block_record(block_digest).await;
+
+        let Some(block_record) = maybe_block_record else {
+            // If genesis, get the addition records from there
+            if block_digest == self.genesis_block().hash() {
+                return Some(
+                    self.genesis_block()
+                        .body()
+                        .transaction_kernel()
+                        .outputs
+                        .iter()
+                        .enumerate()
+                        .map(|(i, ar)| (*ar, Some(i as u64)))
+                        .collect::<HashMap<_, _>>(),
+                );
+            }
+
+            // No block record and not genesis block => nothing we can do
+            return None;
+        };
+
+        // In the common case, the block is canonical, and then it is faster to
+        // read the addition records from the archival mutator set.
+        let block_is_canonical = self.block_belongs_to_canonical_chain(block_digest).await;
+        if block_is_canonical {
+            let aocl_leaf_indices = block_record.min_aocl_index..=block_record.max_aocl_index();
+            let addition_records = self
+                .archival_mutator_set
+                .ams()
+                .aocl
+                .get_leaf_range_inclusive_async(aocl_leaf_indices.clone())
+                .await;
+            Some(
+                addition_records
+                    .into_iter()
+                    .map(|digest| AdditionRecord {
+                        canonical_commitment: digest,
+                    })
+                    .zip(aocl_leaf_indices.into_iter().map(Some))
+                    .collect::<HashMap<_, _>>(),
+            )
+        }
+        // If the block is not canonical, we get the addition records from the
+        // block itself.
```

Comment:
Maybe comment here that AOCL leaf indices will be set to `None` bc not canonical?